src/core/service.c

```diff
@@ -768,6 +768,10 @@ static int service_load(Unit *u) {
         if (r < 0)
                 return r;
 
+        r = exec_context_resolve_cpu_affinity(&s->exec_context);
+        if (r < 0)
+                return r;
```

Comment:
please instead of propagating this at load time, just add an accessor which does this when the field is needed. i.e. add exec_context_get_cpu_affinity() which returns c->cpu_set if it is set, and converts from the numa stuff otherwise. And then always use the accessor when we need the affinity.

Reply:
Again, you should be able to specify,

```
; 0-3 -> node0
; 4-7 -> node1

NUMAMask=1
CPUAffinity=0-3`
CPUAffinity="numa"
```
Then the resulting CPUAffinity mask is 0-7. I think we need to resolve this at unit load time because the set of CPUs may change and it would be weird if CPUAffinity mask would change after the unit is loaded.

Reply:
humm, this appears overly complex, no? is anyone actually going to use it that way? this sounds like a mostly theoretic concept of configuration, not something people would actually use no?

it appears to me either people want to hook the affinity to numa or they don't, but manually calculating in your mind what happens if you combine the two in specific ways sounds unnecessary...

Also, I am pretty sure assigning "numa" should also unconditionally reset the the affinity to the exact value of the numa configuration

Reply:
I agree that it is a bit more complex than it needs to be, but CPUAffinity used to work that way and it is documented that work this way. `"numa"` setting is just syntactic sugar (string "numa" is equivalent to some cpu set) to make things nicer to use.

IOW, I think we shouldn't change how `CPUAffinity=` behaves at this point. Also proposed patch does the right thing and can handle all possible combinations and orders of assignments. I also prepped the test case [1] that we might want to have upstream.

[1] https://gist.github.com/msekletar/e8cb9214f59e0cf2c0835780c9414496